sdks/java/io/rrio/src/main/java/org/apache/beam/io/requestresponseio/CallShouldBackoffBasedOnRejectionProbability.java

```diff
@@ -0,0 +1,102 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.io.requestresponseio;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+/** Reports whether to apply backoff based on https://sre.google/sre-book/handling-overload/. */
+class CallShouldBackoffBasedOnRejectionProbability<ResponseT>
+    implements CallShouldBackoff<ResponseT> {
+
+  // Default multiplier value recommended by https://sre.google/sre-book/handling-overload/
+  private static final double DEFAULT_MULTIPLIER = 2.0;
+
+  // The threshold is the value that pReject() must exceed in order to report a value() of true. If
+  // null, then the computation relies on a random value.
+  private @Nullable Double threshold;
+
+  // The multiplier drives the impact of accepts on the rejection probability. See <a
+  // https://sre.google/sre-book/handling-overload/ for details.
+  private final double multiplier;
+
+  // The number of total requests called to a remote API.
+  private double requests = 0;
+
+  // The number of total accepts called to a remote API.
+  private double accepts = 0;
+
+  /** Instantiate class with the {@link #DEFAULT_MULTIPLIER}. */
+  CallShouldBackoffBasedOnRejectionProbability() {
+    this(DEFAULT_MULTIPLIER);
+  }
+
+  /**
+   * Instantiates class with the provided multiplier. The multiplier drives the impact of accepts on
+   * the rejection probability. See https://sre.google/sre-book/handling-overload/ for details.
+   */
+  CallShouldBackoffBasedOnRejectionProbability(double multiplier) {
+    this.multiplier = multiplier;
+  }
+
+  /**
+   * Setter for the threshold that overrides usage of a random value. The threshold is the value
+   * that pReject() must exceed in order to report a value() of true.
+   */
+  CallShouldBackoffBasedOnRejectionProbability<ResponseT> setThreshold(double threshold) {
+    this.threshold = threshold;
+    return this;
+  }
+
+  /** Update the state of whether to backoff using information about the exception. */
+  @Override
+  public void update(UserCodeExecutionException exception) {
+    this.requests++;
+  }
+
+  /** Update the state of whether to backoff using information about the response. */
+  @Override
+  public void update(ResponseT response) {
+    this.requests++;
+    this.accepts++;
+  }
+
+  /** Provide a threshold to evaluate backoff. */
+  double threshold() {
+    if (this.threshold != null) {
+      return this.threshold;
+    }
+    return Math.random();
+  }
+
+  /**
+   * Compute the probability of API call rejection based on
+   * https://sre.google/sre-book/handling-overload/.
+   */
+  double pReject() {
+    double numerator = requests - multiplier * accepts;
+    double denominator = requests + 1;
```

Comment:
I'd rather keep it as is. It looks more readable in my opinion. Also, the threshold method helpful to keep separate as it has the conditional whether someone supplied a hard code for testing or whether it is using the random. The code is simple enough but I wanted to keep the use of Math.random out of this core method. I will go ahead and resolve this but please re-open if you feel strongly otherwise.